In [39]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Test to Try and Fix Feature Importance issues


#### Initialize the Dataframe and Process the Data - Titanic Dataset

In [46]:
# Import titanic data
titanic = pd.read_csv('data/titanic.csv')

# Processing features:
# 1. Drop columns that are not useful
titanic_features = ['Pclass','Sex','Age','SibSp','Parch','Fare','Cabin','Embarked']
titanic_df = titanic[titanic_features]
titanic_df['CabinLetter'] = titanic_df['Cabin'].str.slice(0,1)
X = titanic_df.drop('Cabin',axis=1)
X['CabinLetter'] = X['CabinLetter'].fillna("?")
X['Pclass'] = X['Pclass'].astype(str)
X['SibSp'] = X['SibSp'].astype(str)
X['Parch'] = X['Parch'].astype(str)
X = X.dropna()
t = titanic.loc[X.index, 'Survived']
# 2. Convert categorical features to dummy variables (one-hot encoding with pandas)
X = pd.get_dummies(X)
# Display summary statistics for the features
print(f"These are the summary statistics for the features:\n {X.describe()}")
print("--------------------------------------------------------------------")
print(f"This is the number of rows in the dataset: {len(X)}")
print("--------------------------------------------------------------------")
print(f"This is the number of unique values in each column:\n {X.nunique()}")

These are the summary statistics for the features:
               Age        Fare
count  712.000000  712.000000
mean    29.642093   34.567251
std     14.492933   52.938648
min      0.420000    0.000000
25%     20.000000    8.050000
50%     28.000000   15.645850
75%     38.000000   33.000000
max     80.000000  512.329200
--------------------------------------------------------------------
This is the number of rows in the dataset: 712
--------------------------------------------------------------------
This is the number of unique values in each column:
 Age               88
Fare             219
Pclass_1           2
Pclass_2           2
Pclass_3           2
Sex_female         2
Sex_male           2
SibSp_0            2
SibSp_1            2
SibSp_2            2
SibSp_3            2
SibSp_4            2
SibSp_5            2
Parch_0            2
Parch_1            2
Parch_2            2
Parch_3            2
Parch_4            2
Parch_5            2
Parch_6            2
Embarked_C         2

/var/folders/c_/j4rx3d1d2md8w8cz828b40jh0000gn/T/ipykernel_9181/1195426193.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  titanic_df['CabinLetter'] = titanic_df['Cabin'].str.slice(0,1)


#### Initialize the Dataframe and Process the Data - Divorce Dataset

Note: The divorce dataset has no missing values and there are exactly 5 unique values per feature (0, 1, 2, 3, 4) that correspond to the following:<br>
- 0 - Never
- 1 - Seldom
- 2 - Averagely
- 3 - Frequently
- 4 - Always

In [47]:
# Import divorce data
divorce = pd.read_csv('data/divorce_data.csv',sep=';')

# Processing features:
X = divorce.drop('Divorce',axis=1) # .dropna() - dataset has no missing values
t = divorce['Divorce']

# Display summary statistics for the features
print(f"These are the summary statistics for the features:\n {X.describe()}")
print("--------------------------------------------------------------------")
print(f"This is the number of rows in the dataset: {len(X)}")
print("--------------------------------------------------------------------")
print(f"This is the number of unique values in each column:\n {X.nunique()}")

These are the summary statistics for the features:
                Q1          Q2          Q3          Q4          Q5          Q6  \
count  170.000000  170.000000  170.000000  170.000000  170.000000  170.000000   
mean     1.776471    1.652941    1.764706    1.482353    1.541176    0.747059   
std      1.627257    1.468654    1.415444    1.504327    1.632169    0.904046   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      2.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
75%      3.000000    3.000000    3.000000    3.000000    3.000000    1.000000   
max      4.000000    4.000000    4.000000    4.000000    4.000000    4.000000   

               Q7          Q8          Q9         Q10  ...         Q45  \
count  170.000000  170.000000  170.000000  170.000000  ...  170.000000   
mean     0.494118    1.452941    1.458824    1.576471  ...    2.458824

#### Train-Test Split 

In [48]:
# Done with sklearn
from sklearn.model_selection import train_test_split

# Split the data into training and validation sets
train_X, val_X, train_y, val_y = train_test_split(X, t, test_size=0.33, random_state=42)

# Print the train X and y shapes
print(f"Train X shape: {train_X.shape}")
print(f"Train y shape: {train_y.shape}")

# Print the summary statistics of the training features and testing features
print(f"Train X summary statistics:\n {train_X.describe()}")
print("--------------------------------------------------------------------")
print(f"Test X summary statistics:\n {val_X.describe()}")

Train X shape: (113, 54)
Train y shape: (113,)
Train X summary statistics:
                Q1          Q2          Q3          Q4          Q5          Q6  \
count  113.000000  113.000000  113.000000  113.000000  113.000000  113.000000   
mean     1.725664    1.522124    1.672566    1.442478    1.460177    0.796460   
std      1.604947    1.414828    1.423291    1.511492    1.581339    0.927355   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      2.000000    2.000000    2.000000    1.000000    1.000000    1.000000   
75%      3.000000    3.000000    3.000000    3.000000    3.000000    1.000000   
max      4.000000    4.000000    4.000000    4.000000    4.000000    4.000000   

               Q7          Q8          Q9         Q10  ...         Q45  \
count  113.000000  113.000000  113.000000  113.000000  ...  113.000000   
mean     0.477876    1.336283    1.380531    1

#### KNN Model
Here we'll instantiate and train a KNN model

In [49]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, f1_score

knn = KNeighborsClassifier(n_neighbors=5)

# Fit the model
knn.fit(train_X, train_y)


KNeighborsClassifier()

#### Baseline Accuracy

In [50]:
knn_pred = knn.predict(val_X)
original_accuracy = accuracy_score(val_y, knn_pred)
print(f"The prediction accuracy is: {original_accuracy}")

AttributeError: 'Flags' object has no attribute 'c_contiguous'

#### Does Accuracy Change with New Split? YES!

In [51]:
# Split the data into NEW training and validation sets
train_X2, val_X2, train_y2, val_y2 = train_test_split(X, t, test_size=0.33)
knn_pred2 = knn.predict(val_X2)
new_accuracy = accuracy_score(val_y2, knn_pred2)
print(f"The prediction accuracy is: {new_accuracy}")

AttributeError: 'Flags' object has no attribute 'c_contiguous'

#### Permutation Feature-Importance

In [52]:
npermutations = 10

feature_names = train_X.columns.tolist()
importances = {}
for col in train_X.columns:
    importances[col] = 0

# Loop through the features and make predictions on permuted data
for col in train_X.columns:
    for perm in range(npermutations):
        # Permute the column of the validation data
        val_X_perm = val_X.copy()
        val_X_perm[col] = val_X[col].sample(frac=1, replace=False).values # np.random.permutation(val_X_perm[col])
        # Make predictions on the new data
        preds = knn.predict(val_X_perm)
        # Compute the accuracy score
        permuted_accuracy = accuracy_score(val_y, preds)

        # Calculate feature importance
        importances[col] += original_accuracy - permuted_accuracy
    
    # Normalize importances
    importances[col] /= npermutations

display(importances)

AttributeError: 'Flags' object has no attribute 'c_contiguous'

In [53]:
## Permutation Feature Importance on the KNN Classifier with all 54 features
## (calculated manually - i.e. without sci-kit learn)

# Define feature names
feature_names = features.columns.tolist()

# Convert pandas DataFrames to numpy arrays for faster computation
X_test = features_test.copy()
y_test = target_test.to_numpy()

# Initialize dictionary to store feature importances
importances = {}
for col in Xtrain.columns:
    importances[col] = 0

# Compute the original accuracy

original_accuracy = accuracy_score(y_test.astype(str), target_pred_knn)
"""['0' '1' '0' '1' '0' '0' '0' '1' '0' '1' '1' '0' '0' '1' '1' '0' '1' '0'
 '1' '0' '1' '1' '1' '1' '1' '0' '1' '0' '1' '1' '1' '1' '0' '0']"""
for i in [0]:#range(X_test.shape[1]):
    print(i)
    # Permute the ith feature in a copy of the test set
    X_test_permuted = X_test.copy()
    #print(X_test_permuted[:10, i])
    np.random.shuffle(X_test_permuted[:, i])
    #print(X_test_permuted[:10, i])

    # Make predictions using the permuted data
    #print(type(X_test_permuted))
    predictions_permuted = knn.predict(X_test_permuted)
    #print(predictions_permuted)
    
    # Compute the accuracy
    accuracy_permuted = accuracy_score(y_test.astype(str), predictions_permuted)
    
    # Compute the feature importance
    importance = original_accuracy - accuracy_permuted
    
    # Add the feature importance to the list
    importances.append(importance)

print(importances)
"""
# Create a DataFrame to store the results
perm_importance_custom_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': importances
})

# Sort the DataFrame by importance
perm_importance_custom_df = perm_importance_custom_df.sort_values(by='Importance', ascending=False)

perm_importance_custom_df"""


NameError: name 'Xtrain' is not defined

# Read in the data

In [54]:
## Importing the dataset
data = pd.read_csv('data/divorce_data.csv', sep=';')

#display(data.head())

## display the data type of the columns
#display(data.dtypes)

targetData = data['Divorce']
## Making two copies of the data for shuffling
featureData = data.copy().drop('Divorce', axis=1)
print(f'Length of the original featureData: {len(featureData)}')
# randomly split the dataset into two separate datasets
trainFeatureData = featureData.sample(frac=0.5, random_state=0)
trainTargetData = targetData[featureData1.index]
testFeatureData = featureData.drop(featureData1.index)
testTargetData = targetData[featureData2.index]
print(f'Length of the featureData1: {len(featureData1)}')
print(f'Length of the targetData1: {len(targetData1)}')
print(f'Length of the featureData2: {len(featureData2)}')
print(f'Length of the targetData2: {len(targetData2)}')
# check for missing values in the split datasets and print results
print(f'Number of missing values in featureData1: {featureData1.isnull().sum().sum()}')
print(f'Number of missing values in targetData1: {targetData1.isnull().sum()}')
print(f'Number of missing values in featureData2: {featureData2.isnull().sum().sum()}')
print(f'Number of missing values in targetData2: {targetData2.isnull().sum()}')

# print the indices of the featureData1 and featureData2
print(f'featureData1 indices:\n {featureData1.index}')
print(f'featureData2 indices:\n {featureData2.index}')

# shuffle the indices of the featureData1 and featureData2
featureData1 = featureData1.sample(frac=1, random_state=42)
featureData2 = featureData2.sample(frac=1, random_state=42)

#print(featureData1.dtypes)
#featureData1.describe()


Length of the original featureData: 170


NameError: name 'featureData1' is not defined

In [55]:
## Testing the data (our two splits)
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, f1_score

knn = KNeighborsClassifier(n_neighbors=3)

shuffledFeatures = featureData.sample(frac=1)#, random_state=42)
#shuffledTargets = targetData.sample(frac=1, random_state=42)

# Fit the model to the training data
knn.fit(featureData, targetData)

# Predict the target for the test data
target_pred_knn = knn.predict(shuffledFeatures)

# Compute accuracy
accuracy_knn = accuracy_score(targetData, target_pred_knn)

# Compute F1 score
f1_knn = f1_score(targetData, target_pred_knn)

accuracy_knn, f1_knn


## When testing with targetData2 and featureData2 we get an accuracy of 0.97647 and an F1 score of 0.97826
## - 

## When testing with targetData1 and featureData1 we get an accuracy of 0.97647 and an F1 score of 0.97222
## - NOW WITH SHUFFLED DATA: .5294117647 & 0.545454



AttributeError: 'Flags' object has no attribute 'c_contiguous'

# Data Exploration

In [56]:
# Get the shape of the dataset
num_rows, num_cols = data.shape

# Check for missing values
missing_values = data.isnull().sum().sum()

# Check the balance of the target variable
divorce_counts = data['Divorce'].value_counts()

num_rows, num_cols, missing_values, divorce_counts


(170,
 55,
 0,
 Divorce
 0    86
 1    84
 Name: count, dtype: int64)

The dataset contains 170 rows (i.e., couples) and 55 columns (54 predictors and 1 target). There are no missing values in the dataset, which is good as it simplifies the preprocessing steps.

The target variable "Divorce" is fairly balanced with 86 instances of non-divorced couples (value 0) and 84 instances of divorced couples (value 1). This is beneficial because imbalanced datasets can often lead to biased models.

# Data Preprocessing

In [57]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Separate features and target
features = data.drop('Divorce', axis=1)#.iloc[:, :5]
target = data['Divorce']

# Split the data into training and test sets
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=42)

# Standardize features
scaler = StandardScaler()
features_train = scaler.fit_transform(features_train)
features_test = scaler.transform(features_test)

features_train.shape, features_test.shape, target_train.shape, target_test.shape


((136, 54), (34, 54), (136,), (34,))

The data has been successfully split into training and test sets. We have 136 instances in the training set and 34 instances in the test set. Each instance has 54 features.

In [58]:
target_train.astype(str)

69     1
138    0
2      1
93     0
136    0
      ..
71     1
106    0
14     1
92     0
102    0
Name: Divorce, Length: 136, dtype: object

In [59]:
target_test.astype(str).values

array(['0', '1', '0', '1', '0', '0', '0', '1', '0', '1', '1', '0', '0',
       '1', '1', '0', '1', '0', '1', '0', '1', '1', '1', '1', '1', '1',
       '1', '0', '1', '1', '1', '1', '0', '0'], dtype=object)

In [60]:
target_pred_knn

array(['0', '1', '0', '1', '0', '0', '0', '1', '0', '1', '1', '0', '0',
       '1', '1', '0', '1', '0', '1', '0', '1', '1', '1', '1', '1', '0',
       '1', '0', '1', '1', '1', '1', '0', '0'], dtype=object)

In [61]:
features_train.shape

(136, 54)

In [62]:
## KNN Classifier using all 54 features in the dataset

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, f1_score

knn = KNeighborsClassifier(n_neighbors=1)

# Fit the model to the training data
knn.fit(features_train.astype(float), target_train.astype(str))

# Predict the target for the test data
target_pred_knn = knn.predict(features_test.astype(float))

# Compute accuracy
accuracy_knn = accuracy_score(target_test.astype(str), target_pred_knn)

# Compute F1 score
f1_knn = f1_score(target_test.astype(str), target_pred_knn,pos_label='1')

accuracy_knn, f1_knn

(0.9705882352941176, 0.9743589743589743)

In [63]:
X_test_permuted

NameError: name 'X_test_permuted' is not defined

In [64]:
y_test

array([0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0])

In [65]:
## Permutation Feature Importance on the RandomForest Classifier with all 54 features
## - note: this may be better than KNN initially with lots of features

from sklearn.ensemble import RandomForestClassifier
from sklearn.inspection import permutation_importance

# Initialize the RandomForestClassifier
rf = RandomForestClassifier(random_state=42)

# Fit the model to the training data
rf.fit(features_train, target_train)

# Compute permutation feature importance
perm_importance = permutation_importance(rf, features_test, target_test, n_repeats=10, random_state=42)

importances = rf.feature_importances_

# Create a DataFrame to store the results
perm_importance_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': perm_importance['importances_mean']
})

# Sort the DataFrame by importance
perm_importance_df = perm_importance_df.sort_values(by='Importance', ascending=False)

perm_importance_df

Feature  Importance
0       Q1         0.0
40     Q41         0.0
29     Q30         0.0
30     Q31         0.0
31     Q32         0.0
32     Q33         0.0
33     Q34         0.0
34     Q35         0.0
35     Q36         0.0
36     Q37         0.0
37     Q38         0.0
38     Q39         0.0
39     Q40         0.0
41     Q42         0.0
1       Q2         0.0
42     Q43         0.0
43     Q44         0.0
44     Q45         0.0
45     Q46         0.0
46     Q47         0.0
47     Q48         0.0
48     Q49         0.0
49     Q50         0.0
50     Q51         0.0
51     Q52         0.0
52     Q53         0.0
28     Q29         0.0
27     Q28         0.0
26     Q27         0.0
25     Q26         0.0
2       Q3         0.0
3       Q4         0.0
4       Q5         0.0
5       Q6         0.0
6       Q7         0.0
7       Q8         0.0
8       Q9         0.0
9      Q10         0.0
10     Q11         0.0
11     Q12         0.0
12     Q13         0.0
13     Q14         0.0
14     Q15         0.0
15     Q16         0.0
16     Q17         0.0
17     Q18         0.0
18     Q19         0.0
19     Q20         0.0
20     Q21         0.0
21     Q22         0.0
22     Q23         0.0
23     Q24         0.0
24     Q25         0.0
53     Q54         0.0

# Feature Importance

In this step, we'll use a decision tree-based method to rank the importance of the features in predicting divorce. This will help us identify the key predictors of divorce.

We'll use the Random Forest algorithm from scikit-learn for this. A Random Forest is an ensemble of Decision Trees that is often used for feature selection because it provides a measure of the importance of each feature.

In [66]:
from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest classifier
rf = RandomForestClassifier(random_state=42)

# Fit the model to the training data
rf.fit(features_train, target_train)

# Get feature importances
importances = rf.feature_importances_

# Create a DataFrame of features and importances
feature_importances = pd.DataFrame({
    'Feature': features.columns,
    'Importance': importances
})

# Sort the DataFrame by importance
feature_importances = feature_importances.sort_values(by='Importance', ascending=False)

display(feature_importances)

# Select the top 10 features
top_features = feature_importances['Feature'][:10].tolist()

# Select these top features from the training and test data
features_train_selected = features_train[top_features]
features_test_selected = features_test[top_features]

features_train_selected.head()


Feature  Importance
39     Q40    0.096593
16     Q17    0.095148
17     Q18    0.091988
18     Q19    0.089627
11     Q12    0.089565
19     Q20    0.063145
15     Q16    0.057146
10     Q11    0.055608
14     Q15    0.047422
25     Q26    0.041697
35     Q36    0.038958
24     Q25    0.028858
29     Q30    0.026860
38     Q39    0.022693
28     Q29    0.020050
40     Q41    0.018753
27     Q28    0.017857
43     Q44    0.016946
8       Q9    0.010794
9      Q10    0.010230
3       Q4    0.009985
37     Q38    0.009709
7       Q8    0.008121
32     Q33    0.003204
5       Q6    0.002893
48     Q49    0.002847
2       Q3    0.002497
0       Q1    0.002424
1       Q2    0.001888
20     Q21    0.001837
4       Q5    0.001617
41     Q42    0.001600
26     Q27    0.001591
36     Q37    0.001242
51     Q52    0.001167
31     Q32    0.001150
33     Q34    0.001019
12     Q13    0.000961
30     Q31    0.000757
52     Q53    0.000703
50     Q51    0.000614
46     Q47    0.000577
44     Q45    0.000484
47     Q48    0.000443
45     Q46    0.000401
53     Q54    0.000197
49     Q50    0.000136
23     Q24    0.000000
42     Q43    0.000000
13     Q14    0.000000
21     Q22    0.000000
22     Q23    0.000000
34     Q35    0.000000
6       Q7    0.000000

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

The Random Forest has ranked the features by their importance in predicting the target variable "Divorce".

The five most important features, according to this model, are:

Q40 with an importance of approximately 0.0966

Q17 with an importance of approximately 0.0951

Q18 with an importance of approximately 0.0920

Q19 with an importance of approximately 0.0896

Q12 with an importance of approximately 0.0896

These results suggest that these questions may be particularly important in predicting divorce.

# Feature Selection

As a starting point, let's choose the top 10 features. However, we can adjust this number later if necessary. Now, let's select these top features from our training and test datasets.

In [ ]:
# Select the top 10 features
top_features = feature_importances['Feature'][:10].tolist()

# Select these top features from the training and test data
features_train_selected = features_train[top_features]
features_test_selected = features_test[top_features]

features_train_selected.head()


Q40  Q17  Q18  Q19  Q12  Q20  Q16  Q11  Q15  Q26
69     0    4    4    4    4    4    4    4    4    4
138    0    0    0    0    0    0    0    0    0    0
2      3    3    3    3    4    2    3    3    3    2
93     0    0    0    0    0    0    0    0    0    0
136    0    1    0    0    1    0    0    0    0    0

# Implementing k-Nearest Neighbors with Scikt Learn

To train and evaluate a k-NN model using scikit-learn, we can use the KNeighborsClassifier class. After training the model, we can use it to make predictions on the test set, and then compute accuracy and F1 score. We are only going to use the top 10 features.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, f1_score

# Initialize the KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=3)

# Fit the model to the training data
knn.fit(features_train_selected, target_train)

print(features_test_selected)

# Predict the target for the test data
target_pred_knn = knn.predict(features_test_selected)

# Compute accuracy
accuracy_knn = accuracy_score(target_test, target_pred_knn)

# Compute F1 score
f1_knn = f1_score(target_test, target_pred_knn)

accuracy_knn, f1_knn


NameError: name 'features_train_selected' is not defined

The k-Nearest Neighbors (k-NN) model from scikit-learn achieved an accuracy of approximately 0.971 (or 97.1%) and an F1 score of approximately 0.974 on the test data.

# Implementing k-Nearest Neighbors from Scratch

Now let's move on to implement the k-NN algorithm from scratch. The steps are as follows:

1.) Calculate Euclidean distance between two instances.


2.) Get the k nearest neighbors of a given test instance.

3.) Predict the class of the test instance by taking the mode of the class labels of the k nearest neighbors.

Let's start by defining the function for calculating the Euclidean distance.

Converting our DataFrames to numpy arrays will make life easier!

In [72]:

X_train = features_train.to_numpy()
y_train = target_train.to_numpy()
X_test = features_test.to_numpy()
y_test = target_test.to_numpy()

print("Data before conversion to Numpy arrays:")
print("features_train:")
print(features_train.head())
print("target_train:")
print(target_train.head())
print("features_test:")
print(features_test.head())
print("target_test:")
print(target_test.head())

print("\nData after conversion to Numpy arrays:")
print("X_train:")
print(X_train)
print("y_train:")
print(y_train)
print("X_test:")
print(X_test)
print("y_test:")
print(y_test)

AttributeError: 'numpy.ndarray' object has no attribute 'to_numpy'

In [67]:
def calculate_euclidean_distance(instance1, instance2):
    """
    Calculate the Euclidean distance between two instances.
    - instance1: first instance
    - instance2: second instance
    """
    return np.sqrt(np.sum((instance1 - instance2) ** 2)) # subtracts corresponding elements of the two arrays

print(f"Euclidean distance between the first and second training instance: {calculate_euclidean_distance(X_train[0], X_train[1])}")


NameError: name 'X_train' is not defined

Next, we'll implement the function to get the k nearest neighbors of a given test instance. This function will compute the Euclidean distance from the test instance to each training instance, keep track of the k instances with the smallest distances, and return their indices.

In [68]:
def get_k_nearest_neighbors(X_train, x_test, k):
    """
    Get the k nearest neighbors of a test instance.
    - X_train: training features
    - x_test: test instance
    - k: number of neighbors to return
    """
    # Calculate the Euclidean distance from the test instance to each training instance
    distances = np.array([calculate_euclidean_distance(x_train, x_test) for x_train in X_train])
    
    # Get the indices of the k training instances with the smallest distances
    nearest_neighbors = distances.argsort()[:k]
    
    return nearest_neighbors

# Test the function
print(f"Indices of the 3 nearest neighbors of the first test instance: {get_k_nearest_neighbors(X_train, X_test[0], 3)}")


NameError: name 'X_train' is not defined

Now, we implement the function to predict the class of a test instance. This function will get the k nearest neighbors of the test instance, find the most common class label among these neighbors, and return this class label as the prediction.

In [69]:
from scipy import stats
def predict_with_k_nearest_neighbors(X_train, y_train, x_test, k):
    """
    Predict the class of a test instance using the k nearest neighbors.
    - X_train: training features
    - y_train: training target values
    - x_test: test instance
    - k: number of neighbors to consider
    """
    # Get the k nearest neighbors of the test instance
    nearest_neighbors = get_k_nearest_neighbors(X_train, x_test, k)
    
    # Get the class labels of the nearest neighbors
    class_labels = y_train[nearest_neighbors]
   
    # Predict the most common class label
    prediction = stats.mode(class_labels)[0]
    
    return prediction

# Test the function
prediction = predict_with_k_nearest_neighbors(X_train, y_train, X_test[0], 3)
print(f"Predicted class: {prediction}, Actual class: {target_test.values[0]}")


NameError: name 'X_train' is not defined

The next step is to use this function to make predictions for multiple test instances. 

In [70]:
def predict_with_k_nearest_neighbors_multiple(X_train, y_train, X_test, k):
    """
    Predict the class of multiple test instances using the k nearest neighbors.
    - X_train: training features
    - y_train: training target values
    - X_test: test features
    - k: number of neighbors to consider
    """
    # Make predictions for each test instance
    predictions = [predict_with_k_nearest_neighbors(X_train, y_train, x_test, k) for x_test in X_test]
    
    return predictions

# Test the function
predictions = predict_with_k_nearest_neighbors_multiple(X_train, y_train, X_test, 3)
print(f"Predicted classes: {predictions[:10]}, Actual classes: {y_test[:10]}")


NameError: name 'X_train' is not defined

The custom implementation of the k-Nearest Neighbors (k-NN) algorithm is working correctly. It made correct predictions for the first 10 instances in the test set!


Now that we have implemented and tested the k-NN algorithm from scratch, let's evaluate its performance on the entire test set. We'll compute the accuracy and F1 score as we did before.

In [71]:
# Make predictions for the entire test set
predictions = predict_with_k_nearest_neighbors_multiple(X_train, y_train, X_test, 3)

# Compute accuracy
accuracy_knn_custom = accuracy_score(y_test, predictions)

# Compute F1 score
f1_knn_custom = f1_score(y_test, predictions)

accuracy_knn_custom, f1_knn_custom

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)


NameError: name 'X_train' is not defined

The custom implementation of the k-Nearest Neighbors (k-NN) algorithm achieved an accuracy of approximately 0.971 (or 97.1%) and an F1 score of approximately 0.974 on the test data. These values are quite high and turned out to be the same as the scikit-learn k-NN. Now, let's compute the permutation feature importance for both the scikit-learn and custom k-NN models.

# Permutation Feature Importance

In [ ]:
## Importing the dataset
data = pd.read_csv('data/divorce_data.csv', sep=';')
display(data)

from sklearn.model_selection import train_test_split

# Separate features and target
features = data.drop('Divorce', axis=1)
target = data['Divorce']

# Split the data into training and test sets
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=42)

from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest classifier
rf = RandomForestClassifier(random_state=42)
# Fit the model to the training data
rf.fit(features_train, target_train)
# Evaluate the model on the test data
accuracy_rf = rf.score(features_test, target_test)
print(f"Accuracy: {accuracy_rf}")


# Let's start with scikit learn's implementation
from sklearn.inspection import permutation_importance

display(features)
display(target)

# Compute permutation feature importance
result = permutation_importance(rf, features, target, n_repeats=30, random_state=42)

importances = result['importances_mean']

for feature_name, importance in zip(features_test.columns, importances):
    print(f"{feature_name}: {importance}")

from sklearn.model_selection import cross_val_score

# # Create a new RandomForestClassifier
# rf = RandomForestClassifier(random_state=42)
# # Fit the model to the training data
# rf.fit(features_train, target_train)
# # Evaluate the model on the test data
# accuracy_rf = rf.score(features_test, target_test)
# print(f"Accuracy: {accuracy_rf}")

# # Perform cross-validation and compute permutation feature importance
# cv_results = cross_val_score(rf, features, target, cv=5, scoring='accuracy')
# print("Cross-validation Accuracy:", cv_results.mean())

# # Compute permutation feature importance on the entire dataset
# result = permutation_importance(rf, features, target, n_repeats=30, random_state=42)
# importances = result['importances_mean']

# # Print feature importances
# for feature_name, importance in zip(features.columns, importances):
#     print(f"{feature_name}: {importance}")



Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  Q46  Q47  Q48  Q49  Q50  \
0     2   2   4   1   0   0   0   0   0    0  ...    2    1    3    3    3   
1     4   4   4   4   4   0   0   4   4    4  ...    2    2    3    4    4   
2     2   2   2   2   1   3   2   1   1    2  ...    3    2    3    1    1   
3     3   2   3   2   3   3   3   3   3    3  ...    2    2    3    3    3   
4     2   2   1   1   1   1   0   0   0    0  ...    2    1    2    3    2   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...  ...  ...  ...  ...  ...   
165   0   0   0   0   0   0   0   0   0    0  ...    1    0    4    1    1   
166   0   0   0   0   0   0   0   0   0    0  ...    4    1    2    2    2   
167   1   1   0   0   0   0   0   0   0    1  ...    3    0    2    0    1   
168   0   0   0   0   0   0   0   0   0    0  ...    3    3    2    2    3   
169   0   0   0   0   0   0   0   1   0    0  ...    3    4    4    0    1   

     Q51  Q52  Q53  Q54  Divorce  
0      2    3    2    1        1  
1      4    4    2    2        1  
2      1    2    2    2        1  
3      3    2    2    2        1  
4      2    2    1    0        1  
..   ...  ...  ...  ...      ...  
165    4    2    2    2        0  
166    2    3    2    2        0  
167    1    3    0    0        0  
168    2    4    3    1        0  
169    3    3    3    1        0  

[170 rows x 55 columns]

Accuracy: 0.9705882352941176


Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  Q45  Q46  Q47  Q48  Q49  \
0     2   2   4   1   0   0   0   0   0    0  ...    3    2    1    3    3   
1     4   4   4   4   4   0   0   4   4    4  ...    2    2    2    3    4   
2     2   2   2   2   1   3   2   1   1    2  ...    2    3    2    3    1   
3     3   2   3   2   3   3   3   3   3    3  ...    3    2    2    3    3   
4     2   2   1   1   1   1   0   0   0    0  ...    2    2    1    2    3   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...  ...  ...  ...  ...  ...   
165   0   0   0   0   0   0   0   0   0    0  ...    0    1    0    4    1   
166   0   0   0   0   0   0   0   0   0    0  ...    3    4    1    2    2   
167   1   1   0   0   0   0   0   0   0    1  ...    2    3    0    2    0   
168   0   0   0   0   0   0   0   0   0    0  ...    4    3    3    2    2   
169   0   0   0   0   0   0   0   1   0    0  ...    1    3    4    4    0   

     Q50  Q51  Q52  Q53  Q54  
0      3    2    3    2    1  
1      4    4    4    2    2  
2      1    1    2    2    2  
3      3    3    2    2    2  
4      2    2    2    1    0  
..   ...  ...  ...  ...  ...  
165    1    4    2    2    2  
166    2    2    3    2    2  
167    1    1    3    0    0  
168    3    2    4    3    1  
169    1    3    3    3    1  

[170 rows x 54 columns]

0      1
1      1
2      1
3      1
4      1
      ..
165    0
166    0
167    0
168    0
169    0
Name: Divorce, Length: 170, dtype: int64

Q1: 0.0
Q2: 0.0
Q3: 0.0
Q4: 0.0
Q5: 0.0
Q6: 0.0
Q7: 0.0
Q8: 0.0
Q9: 0.0
Q10: 0.0
Q11: 0.0
Q12: 0.0
Q13: 0.0
Q14: 0.0
Q15: 0.0
Q16: 0.0
Q17: 0.0
Q18: 0.0
Q19: 0.0
Q20: 0.0
Q21: 0.0
Q22: 0.0
Q23: 0.0
Q24: 0.0
Q25: 0.0
Q26: 0.0
Q27: 0.0
Q28: 0.0
Q29: 0.0
Q30: 0.0
Q31: 0.0
Q32: 0.0
Q33: 0.0
Q34: 0.0
Q35: 0.0
Q36: 0.0
Q37: 0.0
Q38: 0.0
Q39: 0.0
Q40: 0.0021568627450980317
Q41: 0.0
Q42: 0.0
Q43: 0.0
Q44: 0.0
Q45: 0.0
Q46: 0.0
Q47: 0.0
Q48: 0.0
Q49: 0.0
Q50: 0.0
Q51: 0.0
Q52: 0.0
Q53: 0.0
Q54: 0.0


# Divorce Predictor

In [ ]:
print("Welcome to the Divorce Predictor. Please answer the following questions about your relationship.")
print("Your responses should be one of the following: 0 (Never), 1 (Seldom), 2 (Averagely), 3 (Frequently), 4 (Always)")
print("Please respond as honestly as possible.\n")

questions_df = pd.read_csv('data/reference.tsv', delimiter='|', header=None, names=['attribute_id', 'description'])

selected_feature_names = ['Q40', 'Q17', 'Q18', 'Q19', 'Q12', 'Q20', 'Q16', 'Q11', 'Q15', 'Q26']


# Correct the attribute IDs and re-extract the selected questions
corrected_ids = [str(int(x[1:])) for x in selected_feature_names]
selected_questions = questions_df[questions_df['attribute_id'].isin(corrected_ids)]

# Get the questions in a list
selected_questions_list = selected_questions['description'].tolist()

# Display the selected questions
selected_questions_list

# List of top 10 most important questions
questions = selected_questions_list  # replace with the list of the top 10 questions

# List to store the user's responses
responses = []

# Ask the user to input their responses
for i, question in enumerate(questions, start=1):
    print(f"Question {i}: {question}")
    response = input("Your response (0-4): ")
    while not response.isdigit() or int(response) not in range(5):
        print("Invalid response. Please enter a number from 0 to 4.")
        response = input("Your response (0-4): ")
    responses.append(int(response))

print("\nThank you for your responses. The model is now making a prediction...\n")

# Convert the responses into a format that can be used for prediction
responses_df = pd.DataFrame([responses], columns=selected_feature_names)

# Use the trained model to make a prediction
prediction = knn.predict(responses_df)

# Display the prediction
if prediction == 0:
    print("The model predicts that the couple will not get divorced.")
else:
    print("The model predicts that the couple will get divorced.")

Welcome to the Divorce Predictor. Please answer the following questions about your relationship.
Your responses should be one of the following: 0 (Never), 1 (Seldom), 2 (Averagely), 3 (Frequently), 4 (Always)
Please respond as honestly as possible.

Question 1: I think that one day in the future, when I look back, I see that my spouse and I have been in harmony with each other.
Question 2: My spouse and I have similar values in terms of personal freedom.
Question 3: Our dreams with my spouse are similar and harmonious.
Question 4: We're compatible with my spouse about what love should be.
Question 5: We share the same views about being happy in our life with my spouse
Question 6: My spouse and I have similar ideas about how marriage should be
Question 7: My spouse and I have similar ideas about how roles should be in marriage
Question 8: My spouse and I have similar values in trust.
Question 9: I know my spouse's basic anxieties.
Question 10: We're just starting a discussion before I k